openedx/core/djangoapps/external_user_ids/signals.py

```diff
@@ -0,0 +1,40 @@
+"""
+Signal Handlers for External User Ids to be created and maintainer
+"""
+
+from logging import getLogger
+
+from django.db.models.signals import post_save
+from django.dispatch import receiver
+from student.models import CourseEnrollment
+
+from openedx.core.djangoapps.catalog.utils import get_programs
+from .models import ExternalId, ExternalIdType
+
+LOGGER = getLogger(__name__)
+
+
+@receiver(post_save, sender=CourseEnrollment)
+def create_external_id_for_microbachelors_program(
+    sender, instance, created, **kwargs  # pylint: disable=unused-argument
+):
+    """
+    Watches for post_save signal for creates on the CourseEnrollment table.
+    Generate an External ID if the Enrollment is in a MicroBachelors Program
+    """
+    if (
+        created and
+        instance.user and
+        not ExternalId.user_has_external_id(
```

Comment:
I see now you didn't need to check the previous existence in the `add_new_user_id` method. I still think setting the requirements at the DB level is safer though so we don't really on callers to know to check before they create. Could possibly allow us to drop the `user_has_external_id` check as well, but that's optional.

Reply:
The benefit of this check is that I do not perform the other actions later, like digging into the cache and trying to create the external ID.  I will look at the code again though this is all a good point.

Reply:
I kept this check for the reasons listed, but I did also add the uniqueness constraint to the DB model.